roadmap.js

```diff
@@ -0,0 +1,361 @@
+    // Using unpkg will cause ~50 extra requests.
+    popperPromise = import('https://esm.sh/@popperjs/core@2');
+
+  let { features, browsers } = await fetch('/features.json', {
+    credentials: 'include', // https://stackoverflow.com/a/63814972
+    mode: 'no-cors'
+  }).then(res => res.json());
+
+  let scrollbox = document.getElementById('feature-support-scrollbox');
+  let table = document.getElementById('feature-support');
+
+  function h(name, props = {}, children = []) {
+    let node = Object.assign(document.createElement(name), props);
+    node.append(...children);
+    return node;
+  }
+
+  let tBody = document.createElement('tbody');
+
+  table.append(
+    h('thead', {}, [
+      h('tr', {}, [
+        h('th', { id: 'table-blank' }),
+        h('th', { scope: 'col', id: idMap['table-col']('Your browser') }, ['Your browser']),
+        ...Object.entries(browsers).map(([name, { url, logo }]) =>
+          h('th', { scope: 'col', id: idMap['table-col'](name) }, [
+            h('a', { href: url }, [
+              // https://www.w3.org/WAI/WCAG22/Techniques/html/H2
+              h('img', { src: logo, width: 32, height: 32, alt: '' }),
+              h('br'),
+              name
+            ])
+          ])
+        )
+      ])
+    ]),
+    tBody
+  );
+
+  let featureGroups = partitionArray(
+    Object.entries(features).map(([name, feature]) => Object.assign(feature, { name })),
+    feature => feature.phase >= 4
+  );
+
+  featureGroups = [
+    { name: 'Standardized features', features: featureGroups.matched },
+    { name: 'In-progress proposals', features: featureGroups.unmatched },
+  ];
+
+  // Collect all notes and assign an index to each unique item
+  // { "First unique note": 0, "Second unique note": 1, ...}
+  const notes = Object.values(browsers).flatMap(b =>
+    Object.values(b.features)
+      .filter(s => Array.isArray(s))
+      .map(s => s[1])
+  );
+  const note2index = new Map();
+  let noteIndex = 0;
+  for (const note of notes) {
+    if (!note2index.has(note)) {
+      note2index.set(note, noteIndex++);
+    }
+  }
+
+  // Generate the footnote list. They are later referenced in the actual table.
+  const noteList = document.createElement('ol');
+  // Place footnote list outside of the scolling area
+  scrollbox.parentNode.insertBefore(noteList, scrollbox.nextSibling);
+  for (const [note, index] of note2index) {
+    const item = h('li', { id: `feature-note-${index}` });
+    item.innerHTML = renderNote(note);
+    noteList.appendChild(item);
+  }
+
+  // Create an <a> element that links to the specified footnote.
+  // Also returns the HTML id of the footnote it refers to.
+  function createNoteRef(index) {
+    const id = `feature-note-${index}`;
+    return [id, h('a', { href: `#${id}` }, [`[${toAlphabet(index)}]`])];
+  }
+
+  const columnCount = 2 + Object.keys(browsers).length;
+
+  // Correspond to CSS class names.
+  const STYLE_MAP = {
```

Comment:
As far as I can tell, this is always used together with `icon(...)` call and always to create a `<div class="feature-cell (icon class)">(text)</div>`. Let's merge them into one reusable function. Also, instead of an object, you can use `icon-${name}` template.

Reply:
Good idea! It's a lot cleaner now.